server/sonar-server/src/main/java/org/sonar/server/measure/ws/ComponentTreeAction.java

```diff
@@ -0,0 +1,219 @@
+  static final Set<String> STRATEGIES = ImmutableSortedSet.of(ALL_STRATEGY, CHILDREN_STRATEGY, LEAVES_STRATEGY);
+  static final String NAME_SORT = "name";
+  static final String PATH_SORT = "path";
+  static final String QUALIFIER_SORT = "qualifier";
+  static final String METRIC_SORT = "metric";
+  static final Set<String> SORTS = ImmutableSortedSet.of(NAME_SORT, PATH_SORT, QUALIFIER_SORT, METRIC_SORT);
+  static final String ADDITIONAL_METRICS = "metrics";
+  static final String ADDITIONAL_PERIODS = "periods";
+  static final Set<String> ADDITIONAL_FIELDS = ImmutableSortedSet.of(ADDITIONAL_METRICS, ADDITIONAL_PERIODS);
+
+  private final ComponentTreeDataLoader dataLoader;
+  private final UserSession userSession;
+  private final I18n i18n;
+  private final ResourceTypes resourceTypes;
+
+  public ComponentTreeAction(ComponentTreeDataLoader dataLoader, UserSession userSession, I18n i18n,
+    ResourceTypes resourceTypes) {
+    this.dataLoader = dataLoader;
+    this.userSession = userSession;
+    this.i18n = i18n;
+    this.resourceTypes = resourceTypes;
+  }
+
+  @Override
+  public void define(WebService.NewController context) {
+    WebService.NewAction action = context.createAction(ACTION_COMPONENT_TREE)
+      .setDescription(format("Navigate through components based on the chosen strategy with specified measures. The %s or the %s parameter must be provided.<br>" +
+        "Requires one of the following permissions:" +
+        "<ul>" +
+        "<li>'Administer System'</li>" +
+        "<li>'Administer' rights on the specified project</li>" +
+        "<li>'Browse' on the specified project</li>" +
+        "</ul>" +
+        "When limiting search with the %s parameter, directories are not returned.",
+        PARAM_BASE_COMPONENT_ID, PARAM_BASE_COMPONENT_KEY, Param.TEXT_QUERY))
+      .setResponseExample(getClass().getResource("component_tree-example.json"))
+      .setSince("5.4")
+      .setHandler(this)
+      .addSearchQuery("sonar", "component names", "component keys")
+      .addPagingParams(100, MAX_SIZE)
+      .addMultiSortsParams(SORTS, NAME_SORT, true);
+
+    action.createParam(PARAM_BASE_COMPONENT_ID)
+      .setDescription("Base component id. The search is based on this component. It is not included in the response.")
+      .setExampleValue(UUID_EXAMPLE_02);
+
+    action.createParam(PARAM_BASE_COMPONENT_KEY)
+      .setDescription("Base component key.The search is based on this component. It is not included in the response.")
+      .setExampleValue("org.apache.hbas:hbase");
+
+    action.createParam(PARAM_METRIC_KEYS)
+      .setDescription("Metric keys")
+      .setRequired(true)
+      .setExampleValue("ncloc,complexity,violations");
+
+    action.createParam(PARAM_METRIC_SORT)
+      .setDescription(
+        format("Metric key to sort by. The '%s' param must contains the '%s' value. It must be part of the '%s' parameter", Param.SORT, METRIC_SORT, PARAM_METRIC_KEYS))
```

Comment:
"must contain" without s. Moreover what about: "Key of metric to sort by. The parameter 's' must contain value 'metric'" ?